core-v1/src/main/scala/engine/SimpleCachingEngine.scala

```diff
@@ -0,0 +1,319 @@
+package com.rallyhealth.vapors.v1
+
+package engine
+
+import algebra.{EqualComparable, Expr, Extract, WindowComparable}
+import cats.data.{NonEmptyList, NonEmptyVector}
+import data.ExtractValue.AsBoolean
+import data.{ExprState, FactTable, Window}
+import debug.DebugArgs
+import debug.DebugArgs.Invoker
+import logic.{Conjunction, Disjunction, Negation}
+import cats.{Foldable, Functor}
+
+object SimpleCachingEngine {
+
+  type Key = Any
+  type Result = Any
+
+  type ResultCache = Map[Key, Result]
+
+  final case class CachedResult[+V](
+    value: V,
+    cacheState: ResultCache,
+  )
+
+  class Visitor[OP[_]](
+    protected val factTable: FactTable,
+    protected val resultCache: ResultCache,
+  ) extends Expr.Visitor[Lambda[(`-I`, `+O`) => I => CachedResult[O]], OP]
+    with CommonEngine[OP] {
+
+    import cats.implicits._
+
+    protected def visitWithUpdatedCache[I, O](
+      input: I,
+      expr: Expr[I, O, OP],
+      updated: ResultCache,
+    ): CachedResult[O] = {
+      val visitor = new Visitor[OP](factTable, updated)
+      expr.visit(visitor)(input)
+    }
+
+    protected def visitWithUpdatedCache2[I, A, B, R](
+      input: I,
+      xa: Expr[I, A, OP],
+      xb: Expr[I, B, OP],
+    )(
+      operation: (A, B) => R,
+    ): (A, B, CachedResult[R]) = {
+      val a = xa.visit(this)(input)
+      val b = visitWithUpdatedCache(input, xb, a.cacheState)
+      val r = operation(a.value, b.value)
+      (a.value, b.value, CachedResult(r, b.cacheState))
+    }
+
+    protected def visitWithUpdatedCacheN[I, A](
+      input: I,
+      xs: NonEmptyVector[Expr[I, A, OP]],
+    ): CachedResult[NonEmptyVector[A]] = {
+      val (cacheState, values) = xs.toVector.foldLeft((resultCache, Vector.empty[A])) {
+        case ((updated, values), x) =>
+          val result = visitWithUpdatedCache(input, x, updated)
+          (result.cacheState, values :+ result.value)
+      }
+      // Safe: because the original vector is non-empty and the fold always produces a vector of the same size
+      CachedResult(NonEmptyVector.fromVectorUnsafe(values), cacheState)
+    }
+
+    protected def cached[V](
+      value: V,
+      cacheState: ResultCache = resultCache,
+    ): CachedResult[V] = CachedResult(value, cacheState)
+
+    protected def state[I, O](
+      input: I,
+      output: CachedResult[O],
+    ): ExprState[I, CachedResult[O]] = ExprState(factTable, Some(input), Some(output))
+
+    protected def debugging[E <: Expr.AnyWith[OP]](
+      expr: E,
+    )(implicit
+      debugArgs: DebugArgs[E, OP],
+    ): InvokeAndReturn[E, OP, debugArgs.In, debugArgs.Out] =
+      new InvokeAndReturn(DebugArgs[OP].of(expr)(debugArgs))
+
+    // TODO: Use Hash type bounds on I?
+    protected def memoize[I, O](
+      expr: Expr[I, O, OP],
+      input: I,
+    )(
```

Comment:
This would be the safest option, but it might make the code more complicated.

Reply:
I could potentially store the hash from the previous output (assuming it isn't `ExprState.Nothing`)... I don't think this is really necessary yet though. I might just do a little experiment to prove that the current interface could handle this change in the future.